ethereumj-core/src/main/java/org/ethereum/listener/EventListener.java

```diff
@@ -0,0 +1,304 @@
+     * to the transaction which generated events.
+     * The transaction pending state is tracked via
+     * - onPendingTransactionUpdate callback: to handle Tx state changes on block inclusion, rebranches and rejects
+     * - onBlock callback: to handle confirming blocks after the Tx is included
+     */
+    protected class PendingEvent {
+        // The latest transaction receipt either pending/rejected or best branch included
+        public TransactionReceipt receipt;
+
+        // The Solidity Events (represented as EventData) generated by transaction
+        public List<EventData> eventData;
+        // null if pending/rejected
+        public Block includedTo;
+        // the latest block from the main branch
+        public Block bestConfirmingBlock;
+        // if came from a block we take block time, it pending we take current time
+        public long created;
+        // status of the Ethereum Tx
+        public TxStatus txStatus;
+
+        public PendingEvent(long created) {
+            this.created = created;
+        }
+
+        public void update(TransactionReceipt receipt, List<EventData> txs, PendingTransactionState state, Block includedTo) {
+            this.receipt = receipt;
+            this.eventData = txs;
+            this.bestConfirmingBlock = state == PendingTransactionState.INCLUDED ? includedTo : null;
+            this.includedTo = state == PendingTransactionState.INCLUDED ? includedTo : null;
+            txStatus = state.isPending() ? TxStatus.PENDING :
+                    (state == PendingTransactionState.DROPPED ? TxStatus.REJECTED : TxStatus.getConfirmed(1));
+        }
+
+        public boolean setBestConfirmingBlock(Block bestConfirmingBlock) {
+            if (txStatus == TxStatus.REJECTED || txStatus == TxStatus.PENDING) return false;
+            if (this.bestConfirmingBlock.isEqual(bestConfirmingBlock)) return false;
+            this.bestConfirmingBlock = bestConfirmingBlock;
+            txStatus = TxStatus.getConfirmed((int) (bestConfirmingBlock.getNumber() - includedTo.getNumber() + 1));
+            return true;
+        }
+
+        @Override
+        public String toString() {
+            return "PendingEvent{" +
+                    "eventData=" + eventData +
+                    ", includedTo=" + (includedTo == null ? "null" : includedTo.getShortDescr()) +
+                    ", bestConfirmingBlock=" + (bestConfirmingBlock == null ? "null" : bestConfirmingBlock.getShortDescr()) +
+                    ", created=" + created +
+                    ", txStatus=" + txStatus +
+                    ", tx=" + receipt.getTransaction() +
+                    '}';
+        }
+    }
+
+    protected LogFilter contractAddressFilter;
+    protected CallTransaction.Contract contract;
+    protected PendingStateImpl pendingState;
+
+    // txHash => PendingEvent
+    protected Map<ByteArrayWrapper, PendingEvent> pendings = new LinkedHashMap<>();
```

Comment:
Maybe wrap it by `ByteArrayMap` for convenience?

Reply:
Cool, didn't know about it.